Switch to Pyrra for rhobs metrics SLOs

saswatamcode opened the pull request:
This PR deprecates the slo-libsonnet library usage for rhobs metrics rules, and switches over to Pyrra.

This adds a go file responsible for generating the Prometheus rule files which contain SLO alerts. 

To note, we don't deploy Pyrra, but use its functionality via mimic. It is roughly the same as the pyrra CLI, but customized to fit some of our needs.

TODO:

- Update tests for the rules (or rm completely)
- Update SLOs to include logging SLOs

douglascamata commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,549 @@
+	)
+}
+
+func envSLOs(envName string, objs []pyrrav1alpha1.ServiceLevelObjective, ruleFilename string, isStage bool, genPyrra, genRules *mimic.Generator) {
```
If you would like to improve readability of calls of this function, the `isStage` type could be converted to a typed string too. This makes it obvious whether the env's SLOs are for staging without a person reading/writing the code having to rely on fancy editors features or reading the function definition.

saswatamcode replied:
Yup! Perhaps converting the env names should also be typed. Will look into it.

douglascamata reviewed (commented):
Some initial feedback, nothing major. 👍

saswatamcode commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,549 @@
+)
+
+const (
+	window             = "28d" // Window over which our SLOs are calculated.
+	availabilityTarget = "99"  // Availability SLO target, currently at 99%.
+	latencyTarget      = "90"  // Latency SLO percentile target.
+	latencySeconds     = "5"   // Latency request duration to measure percentile target (this is diff for query SLOs).
+)
```
Yeah, I'll rename, was trying to keep things organized in my head.

saswatamcode replied:
I did rename these, but I still added back the comments, as they seem to help quite a bit on hover, when you want to peek at the definition without jumping. 🙂

saswatamcode commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,549 @@
+	}
+
+	// IDs <> env.
+	dashboardIDs = map[string]string{
+		"telemeter-production":   "f9fa7677fb4a2669f123f9a0f2234b47",
+		"telemeter-staging":      "080e53f245a15445bdf777ae0e66945d",
+		"mst-production":         "283e7002d85c08126681241df2fdb22b",
+		"mst-stage":              "92520ea4d6976f30d1618164e186ef9b",
+		"rhobsp02ue1-production": "7f4df1c2d5518d5c3f2876ca9bb874a8",
+	}
```
I don't think they do change and the IDs/data sources remain stable? We have these templated in jsonnet too btw: https://github.com/rhobs/configuration/blob/main/observability/prometheusrules.jsonnet#L93. I just brought them here so construct the Pyrra configs.

periklis reviewed (commented):
This is really awesome work @saswatamcode :muscle: As mentioned otherwise I feel we are on the right track with pyrra (although it is a toy project) because it closes the loop of uniformly maintain and report SLOS for RHOBS. My two pointers to this PR:
1. Feel free to take over the Loki/Logs SLOS. I have created them as a head start for conversation with the logs tenants. So far the levels are not set in stone. However we can benefit from moving the current discource having a pyrra based reporting.
2. Although pyrra is maintained by two contributors at there free time, it feels right to disconnect from `slo.libsonnet` because it enables co-maintaining the reporting with our AppSRE folks.

saswatamcode commented:
So quick update,

- Added the logs availability SLOs for MST!
- Removed the `observatorium-api-(production|stage).prometheusrules.yaml` SLO alert rules, as they were exact duplicates of the API error budget ones.

periklis commented:
> So quick update,
> 
>     * Added the logs availability SLOs for MST!
>       
>       * cc: @periklis would you want the availability target for this to be the same as metrics, i.e 99% over 28d?

You get set the level to realistically arbitrarily because we don't have the required on-call structure across the two geos. For the start we can leave this to 95% and triage with the log storage team where we stand today.

>       * Also, I noticed there are currently no latency SLOs for logging endpoints. Should they be included?
> 

IIRC the metrics latency SLOs are based on synthetic client metrics (I believe the tool is called avalance). We don't have such a tool support for logs yet. Any suggestions welcome (e.g. using up for this task)?

>     * Removed the `observatorium-api-(production|stage).prometheusrules.yaml` SLO alert rules, as they were exact duplicates of the API error budget ones.
> 
>     * Full deprecated slo-libsonnet

saswatamcode commented:
> You get set the level to realistically arbitrarily because we don't have the required on-call structure across the two geos. For the start we can leave this to 95% and triage with the log storage team where we stand today.

Ack! Will update to 95.

> IIRC the metrics latency SLOs are based on synthetic client metrics (I believe the tool is called avalance). We don't have such a tool support for logs yet. Any suggestions welcome (e.g. using up for this task)?

The query latency is based on the `up` job, yes, which queries metrics written by avalanche to rhobs tenant. The query latencies for metrics rely on # of samples queried in a single request, so 10s for 1M samples, 30s for 10M and so on. 
For write path latency we use all data, i.e, actual tenant data + avalanche combined. 🙂

periklis commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,688 @@
+					"instance": string(envName),
+				},
+				description:         "API logs /query handler is burning too much error budget to guarantee availability SLOs.",
+				successOrErrorsExpr: "http_requests_total{job=\"" + apiJobSelector[envName] + "\", handler=\"query\", group=\"logsv1\", code=~\"^5..$\"}",
+				totalExpr:           "http_requests_total{job=\"" + apiJobSelector[envName] + "\", handler=\"query\", group=\"logsv1\", code!~\"^4..$\"}",
```
In the previous [version](https://github.com/rhobs/configuration/blob/main/resources/observability/prometheusrules/observatorium-api-production.prometheusrules.yaml#L628-L635) based on `slo.libsonnet` I included deliberately the  `label`, `labels` and `label_values` handlers in this one too: `handler=~"query|label|labels|label_values"`

The rationale doing so is that it is less common in Loki to execute instant queries but very regularly tools like Grafana execute `/labels` (Old Grafana 6 calls the deprecated `/label`) and `/label_values` to feed the UI query builder.

Of course for reasons of pure sanity we can add a dedicated SLO for `label`, `labels` and `label_values`. WDYT?

saswatamcode replied:
Yeah, I wasn't sure if we should alert and page on error budget burn of label/label_values call (as this PR adds only alerting function of Pyrra) . Do you still want me to add it?

periklis replied:
Please add them to have them for a start. Errors on labels and label values are a severe case (e.g. index files not available for whatever reason). Latency also affects the entire log access are anything needs to be resolved through the labels first (e.g. which chunks to download from S3).

saswatamcode replied:
Added here https://github.com/rhobs/configuration/pull/412/commits/647754aec88ad2eb1b75430d92a9a0d37f394a6c

periklis reviewed (commented):
LGTM from a Logs SLOS standpoint. One pointer only to improve checking availability for Loki's labels endpoints.

douglascamata commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,702 @@
+			},
+		}
+	case tracesSignal:
+		panic("signal not yet supported")
```
For later, could be a separate PR: this panic message doesn't tell me which signal is not yet supported. Makes it harder to read logs without context on this file.

saswatamcode replied:
Made this + other panics more descriptive

douglascamata commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,702 @@
+	return slos.GetObjectives(envName)
+}
+
+const (
+	yamlExt               = ".yaml"
+	prometheusRuleYamlExt = ".prometheusrules.yaml"
+)
```
Do these really need to be consts? They are very far from where they are used, which makes code reading unnecessarily harder. 

Couldn't we make the `envSLOs` function take "base names", have the extensions there, in the function body, and let the compiler optimize it?

saswatamcode replied:
Removed

douglascamata commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,702 @@
+	case tracesSignal:
+		panic("signal not yet supported")
+	default:
+		panic("not an Observatorium signal")
```
What is not an Observatorium signal? Could we add this information to the panic?

douglascamata commented on configuration/observatorium/slo.go:
```diff
@@ -0,0 +1,702 @@
+	case telemeterProduction, mstProduction, rhobsp02ue1Production:
+		return false
+	default:
+		panic("not a RHOBS env")
```
What is not a RHOBS env? The panic message could indicate it.

saswatamcode commented on Makefile:
```diff
@@ -170,9 +195,18 @@ clean:
-	GOFLAGS="-mod=mod" go run ./mimic.go generate -o resources/.tmp
+	GOFLAGS="-mod=mod" go run ./mimic.go generate -o resources
+
+# In theory we'd be able to run Pyrra as a CLI directly.
+# PYRRA_DIR := $(shell pwd)/test
+#
+# .PHONY: docker-pyrra
+# docker-pyrra:
+# 	@chmod -R 777 $(PYRRA_DIR)
+# 	docker run -v $(PYRRA_DIR):/shared -i ghcr.io/pyrra-dev/pyrra:main generate --config-files=/shared/pyrra/*.yaml --prometheus-folder=/shared/rules/ --generic-rules
```
I'd still keep this around in case we decide to move to pyrra CLI completely at some point. :)

periklis reviewed (commented):
Nice thank you for adding the label endpoints to the Logs read availability SLO.

douglascamata reviewed (approved):
LGTM :shipit: 🚀 🌔

metalmatze commented:
Just stumbled upon this. Wooohooo :tada: 
Let me know upstream or on CNCF Slack if you ever encounter something you want to discuss.

saswatamcode commented:
Thanks for amazing project! 💪🏻

periklis commented:
@metalmatze From the top of head (disclaimer: have not skimmed through thr pyrra repo issues), something like `pyrra serve -f .` for local testing would be awesome :heart: Thanks for this amazing piece!
